x-pack/filebeat/input/awss3/sqs.go

```diff
@@ -47,11 +59,81 @@ func newSQSReader(log *logp.Logger, metrics *inputMetrics, sqs sqsAPI, maxMessag
 	}
 }
 
-func (r *sqsReader) Receive(ctx context.Context) error {
-	// This loop tries to keep the workers busy as much as possible while
+func (r *sqsReader) Receive(ctx context.Context, pipeline beat.Pipeline) error {
+	// The loop tries to keep the ProcessSQS workers busy as much as possible while
 	// honoring the max message cap as opposed to a simpler loop that receives
-	// N messages, waits for them all to finish, then requests N more messages.
-	var workerWg sync.WaitGroup
+	// N messages, waits for them all to finish sending events to the queue, then requests N more messages.
+	var processingWg sync.WaitGroup
+
+	// At the end of the loop, after a shutdown or anything else, we still need to wait for the DeleteSQS goroutines to
+	// complete, otherwise the messages will be sent back to the queue even if they should be deleted.
+	var deletionWg sync.WaitGroup
+
+	// We send to processingChan the outcome of each ProcessSQS call.
+	// We don't want to buffer the processingChan, since it will prevent workers ProcessSQS goroutines to return.
+	processingChan := make(chan processingOutcome)
+
+	// We use deletionChan to throttle the number of DeleteSQS goroutines.
+	// deletionChan := make(chan struct{}, 3200)
+
+	go func() {
+		for {
+			outcome, ok := <-processingChan
+			// processingChang is closed, no more outcomes to process, we can exit.
+			if !ok {
+				return
+			}
+
+			// A ProcessSQS goroutine has sent an outcome, let's process it asynchronously in order to handle SQS message deletion.
+			go func(outcome processingOutcome) {
```

Comment:
Why do you need to start a goroutine for each processing outcome?

The beat pipeline client for each worker already has an event listener receiving a callback every time an event is acked.

https://github.com/elastic/beats/blob/4f46fcb888b7f12706d005b7e1d2d1dafcbe6fcd/x-pack/filebeat/input/awss3/sqs_s3_event.go#L310-L312

This is what is used to implement the awscommon acker that you are waiting on later with `outcome.acker.Wait()`.

Rather than starting a goroutine that blocks on a call to `outcome.acker.Wait()`, why not write an sqs specific acker that calls `DeleteSQS()` when all the events in an s3 object are acknowledged? That way you don't need a goroutine to sit around and wait for that to happen, you can react to the notification from the pipeline.

In the worst case where each s3 object is a single event, you will end up with many individual calls to `DeleteMsg`. Are you able to use [DeleteMessageBatch](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_DeleteMessageBatch.html) to help when multiple SQS messages can be deleted at once? 

This appears to be what vector is using https://github.com/vectordotdev/vector/blob/f34738e6737e79f77dc6aa9aecb8d00430f64d99/src/sources/aws_sqs/source.rs#L204

Reply:
> Rather than starting a goroutine that blocks on a call to `outcome.acker.Wait()`, why not write an sqs specific acker that calls `DeleteSQS()` when all the events in an s3 object are acknowledged? That way you don't need a goroutine to sit around and wait for that to happen, you can react to the notification from the pipeline.

yes, let me give a try and benchmark the scenario

Reply:
> In the worst case where each s3 object is a single event, you will end up with many individual calls to `DeleteMsg`. Are you able to use [DeleteMessageBatch](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_DeleteMessageBatch.html) to help when multiple SQS messages can be deleted at once?

`DeleteMessageBatch` can return `TooManyEntriesInBatchRequest`, and I cannot find yet how much is too many :)
Also if we switch to the specific sqs acker this will have the state of a single message, so we need some other logic to share the states of a batch of sqs ackers to send the `DeleteMessageBatch` call.

Also I'd prefer to avoid handling partial failures coming from the `DeleteMessageBatch` call.

Reply:
> > Rather than starting a goroutine that blocks on a call to `outcome.acker.Wait()`, why not write an sqs specific acker that calls `DeleteSQS()` when all the events in an s3 object are acknowledged? That way you don't need a goroutine to sit around and wait for that to happen, you can react to the notification from the pipeline.
> 
> yes, let me give a try and benchmark the scenario

done, please check the changes.
I will soon post a benchmark comparison, quite promising already.

No more CPU utilization spikes first of all